src/mlpack/methods/reinforcement_learning/q_networks/dueling_dqn.hpp

```diff
@@ -116,7 +118,16 @@ class DuelingDQN
    */
   void Backward(const arma::mat state, arma::mat& target, arma::mat& gradient)
   {
-    featureNetwork.Backward(state, target, gradient);
+    arma::mat gradError;
+    lossFunction.Backward(networkOutput, target, gradError);
+
+    arma::mat gradValue, gradAdvantage;
+    // valueNetwork.Backward(state, gradError, gradValue);
+    advantageNetwork.Backward(features, gradError, gradAdvantage);
+
+    arma::mat gradSum;
+    gradSum = gradAdvantage; // + gradValue
+    featureNetwork.Backward(state, gradSum, gradient);
```

Comment:
@zoq, I tried today to get the gradients from one FNN, and pass to to another FFN.. But I think the gradients returned by the Backward pass are in the form of a flattened matrix. I'm not sure how to proceed

Reply:
The layer takes care of the gradient size, since it knows what the size should look like:

https://github.com/mlpack/mlpack/blob/c890afb5cbccf49e4c7b5f34932aaac78996e1da/src/mlpack/methods/ann/layer/linear_impl.hpp#L80-L83

Not sure maybe I missed something?

Reply:
Oh, I see now.. I think the problem is that `gradAdvantage` gives me the gradients for all the parameters in `advantageNetwork`. But, I think, we'd need to pass just the gradients for the first layer parameters of `advantageNetwork`, into featureNetwork.Backward(). Isn't that correct?

Reply:
That is correct, I think since you know the size you can just use `gradient.submat(...)` you can also use `layer.Gradient()` to get the gradient or iterate over all layers to get the size and use `submat`:

```c++
WeightSizeVisitor weightSizeVisitor;
size_t weights = 0;
size_t weightsLayer = 0;
for (size_t i = 0; i < network.size(); ++i)
{

  if (i == 2)
  {
     weightsLayer = boost::apply_visitor(weightSizeVisitor, network[i]);
     break;
  }
  
  // Get weight size for a single layer until we reached the layer we are interested in.
  weights += boost::apply_visitor(weightSizeVisitor, network[i]);
}

gradient.submat(weights, 0, weightsLayer, 0);
```

Reply:
Well, I think there is one more problem.. I suppose that since we do backward in 'batch mode', so the gradients calculated by the `Backward()` function are a mean of the gradients. eg, if the network(with n parameters) has 10 samples given as input, then we'd want the mean gradients(n in number) for the optimizer, but if we want to pass the gradients to the previous network, we'd want the individual gradients for each sample, and then after passing the gradients back, and before giving the gradients to the optimizer, we can take a mean for the samples.. 
I hope I make sense. :question:

Reply:
In that case, is there a way to avoid all this, and instead specify the branching of a network into two (like here, a feature network branches off to a value and an advantage), at the network definition itself.. 
or maybe a method to concatenate two networks, instead of manually passing forward features in ForwardPass and passing backward gradients in BackwardPass ?

Reply:
I am not sure if I have the right context here but do you mind looking at this [test for concat layer](https://github.com/mlpack/mlpack/blob/68e43d8c80e29e3d4dc16c19cae68cdd9eba7ceb/src/mlpack/tests/ann_layer_test.cpp#L1332). It helps in concatenating two networks.

Reply:
No, @kartikdutt18 you are right, the concat layer is a great idea to solve this.

Reply:
Hi, I'm not sure if we branch a module into two separate modules using `Concat` layer. I think its just for adding two layers in a sequence one after other. Maybe I'm missing something..

I think we'd need something like 
```
module.Add(moduleA, moduleB)
```
instead of 
```
module.Add(moduleA);
module.Add(moduleB);
```

Reply:
Concat, will take the input and run it through all added layers (you can use the sequential layer to add multiple layers) and at the end will take the average over all branches.